File: web/src/types/assets.ts
// Mirrors app/schemas/assets.py.
import type { MapPoint } from "./map";

export interface SolarPlantFields {
  plant_name: string | null;
  location_desc: string | null;
  installed_capacity_mw: string | null;
  interfacing_ss: string | null;
  voltage_level: string | null;
  /** Free text as the legacy form accepted it. */
  commercial_operation_date_raw: string | null;
  division: string | null;
  circle: string | null;
  zone: string | null;
  location: MapPoint | null;
}

export interface SolarPlantOut extends SolarPlantFields {
  solar_id: number;
  commercial_operation_date: string | null;
}

export interface EhvConsumerFields {
  name: string | null;
  location_desc: string | null;
  installed_capacity_mw: string | null;
  feeder_id: number | null;
  feeder_name: string | null;
  substation: string | null;
  consumer_code: string | null;
  voltage_rate: string | null;
  function_loc_code: string | null;
  connected_ss: string | null;
  line_name: string | null;
  line_code: string | null;
  division: string | null;
  circle: string | null;
  zone: string | null;
  location: MapPoint | null;
}

export interface EhvConsumerOut extends EhvConsumerFields {
  ehv_id: number;
}

File: backend/app/schemas/assets.py
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.map import MapPoint

# ---------------------------------------------------------------------------
# Solar plants and EHV consumers. Both are small, flat CRUD tables in the
# legacy system (sp_Solarplant, sp_ehvconsumers) - insert, update, delete,
# then return the whole list. Neither has children or geometry beyond a point.
# ---------------------------------------------------------------------------


class SolarPlantFields(BaseModel):
    plant_name: str | None = None
    location_desc: str | None = None
    installed_capacity_mw: Decimal | None = None
    interfacing_ss: str | None = None
    voltage_level: str | None = Field(default=None, max_length=50)
    # free text as the legacy form had it; parsed server-side where possible
    commercial_operation_date_raw: str | None = Field(default=None, max_length=100)
    division: str | None = Field(default=None, max_length=100)
    circle: str | None = Field(default=None, max_length=100)
    zone: str | None = Field(default=None, max_length=100)
    location: MapPoint | None = None


class SolarPlantOut(SolarPlantFields):
    model_config = ConfigDict(from_attributes=True)

    solar_id: int
    commercial_operation_date: date | None


class EhvConsumerFields(BaseModel):
    name: str | None = None
    location_desc: str | None = None
    installed_capacity_mw: Decimal | None = None
    feeder_id: int | None = None
    feeder_name: str | None = None
    substation: str | None = None
    consumer_code: str | None = Field(default=None, max_length=100)
    voltage_rate: str | None = Field(default=None, max_length=50)
    function_loc_code: str | None = Field(default=None, max_length=100)
    connected_ss: str | None = None
    line_name: str | None = None
    line_code: str | None = Field(default=None, max_length=100)
    division: str | None = Field(default=None, max_length=100)
    circle: str | None = Field(default=None, max_length=100)
    zone: str | None = Field(default=None, max_length=100)
    location: MapPoint | None = None


class EhvConsumerOut(EhvConsumerFields):
    model_config = ConfigDict(from_attributes=True)

    ehv_id: int
